frame/contracts/src/storage/meter.rs

```diff
@@ -0,0 +1,724 @@
+		}
+	}
+
+	/// The amount of balance that is still available from the original `limit`.
+	fn available(&self) -> BalanceOf<T> {
+		self.total_deposit.available(&self.limit)
+	}
+}
+
+/// Functions that only apply to the root state.
+impl<T, E> RawMeter<T, E, Root>
+where
+	T: Config,
+	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
+	E: Ext<T>,
+{
+	/// Create new storage meter for the specified `origin` and `limit`.
+	///
+	/// This tries to [`Ext::check_limit`] on `origin` and fails if this is not possible.
+	pub fn new(
+		origin: &T::AccountId,
+		limit: Option<BalanceOf<T>>,
+		min_leftover: BalanceOf<T>,
+	) -> Result<Self, DispatchError> {
+		let limit = E::check_limit(&origin, limit, min_leftover)?;
+		Ok(Self { limit, ..Default::default() })
+	}
+
+	/// The total amount of deposit that should change hands as result of the execution
+	/// that this meter was passed into.
+	///
+	/// This drops the root meter in order to make sure it is only called when the whole
+	/// execution did finish.
+	pub fn into_deposit(self) -> DepositOf<T> {
+		// Clone is necessary because of the drop implementation.
+		self.total_deposit.clone()
+	}
+}
+
+/// Functions that only apply to the nested state.
+impl<T, E> RawMeter<T, E, Nested>
+where
+	T: Config,
+	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
+	E: Ext<T>,
+{
+	/// Try to charge the `diff` from the meter. Fails if this would exceed the original limit.
+	pub fn charge(&mut self, diff: &Diff) -> Result<DepositOf<T>, DispatchError> {
+		debug_assert!(!self.terminated);
+		let deposit = diff.to_deposit::<T>();
+		let total_deposit = self.total_deposit.saturating_add(&deposit);
+		if let Deposit::Charge(amount) = total_deposit {
+			if amount > self.limit {
+				return Err(<Error<T>>::StorageDepositLimitExhausted.into())
+			}
+		}
+		self.total_deposit = total_deposit;
+		self.own_deposit = self.own_deposit.saturating_add(&deposit);
+		Ok(deposit)
+	}
+
+	/// Charge from `origin` a storage deposit for contract instantiation.
+	///
+	/// This immediatly transfers the balance in order to create the account.
+	pub fn charge_instantiate(
+		&mut self,
+		origin: &T::AccountId,
+		contract: &T::AccountId,
+		info: &mut ContractInfo<T>,
+	) -> Result<DepositOf<T>, DispatchError> {
+		debug_assert!(!self.terminated);
+		let deposit = Diff {
+			bytes_added: info.encoded_size() as u32,
+			items_added: 1,
+			require_ed: true,
+			..Default::default()
+		}
+		.to_deposit::<T>();
+		// We do not increase `own_deposit` because this will be charged later when the contract
+		// execution does conclude.
+		let total_deposit = self.total_deposit.saturating_add(&deposit);
+		if let Deposit::Charge(amount) = &total_deposit {
+			if amount > &self.limit {
+				return Err(<Error<T>>::StorageDepositLimitExhausted.into())
+			}
+		}
+		if let Deposit::Charge(amount) = &deposit {
+			info.storage_deposit = info.storage_deposit.saturating_add(*amount);
```

Comment:
Since this would always be true, could use `StorageDeposit::charge_or_zero` to make it a one liner